README.md
dispersalDiversity
==================


This is a collection of [R](http://www.r-project.org) functions to facilitate
analysis of dispersal in biological communities.  The bulk of the functions
calculate dispersal diversity statistics and allow for comparison of diversity
statistics, as described in Scofield _et al._ 2012 [_American Naturalist_ 180:
719-732](http://www.jstor.org/stable/10.1086/668202).  We originally developed
the diversity tests to understand seed dispersal in plant populations, but the
tests themselves should be useful for biodiversity data or any other
diversity-like data that can be expressed with this same data structure.

There are also new functions for calculating allelic diversity using these same
conceptual and statistical principles, and for comparing allele diversity
statistics.

**NOTE: This version has major changes to the interface.  For the version used in Scofield et al. (2012) with the previous interface, get the [release tagged 0.1](https://github.com/douglasgscofield/dispersalDiversity/releases), but simply `source` the files you are interested in... release 0.1 does not have an R package structure.**

* * *

These statistical tools were developed in collaboration with Peter Smouse ([Rutgers University](http://www.rci.rutgers.edu/~deenr/PES.html)) and  Victoria Sork ([UCLA](https://www.eeb.ucla.edu/Faculty/Sork/Sorklab/)) and were funded by U.S. National Science Foundation awards NSF-DEB-0514956 and NSF-DEB-0516529.

* * *

The **Mann-Whitney-Wilcoxon nested ranks test** we originally provided here has been made an R package:
```R
install.packages("nestedRanksTest")
vignette("nestedRanksTest")
```
and has been moved to [its own repository](https://github.com/douglasgscofield/nestedRanksTest).



Installation
------------

Install from github using `devtools`:

```R
install.packages("devtools")
devtools::install_github("douglasgscofield/dispersalDiversity")
library(dispersalDiversity)
```



Prerequisites
-------------

This package requires the `readGenalex` package v1.0 available from CRAN, which is
a prerequisite for this package and will be installed by default.

For producing plots, many find aesthetic colour choices to be available through
palettes in the `RColorBrewer` package.  `membershipPlot` (see below) will use
this package if it is available.  You may already have it installed, and the
metadata for `dispersalDiversity` lists it as a suggested package.  If it is
not installed, get it from CRAN:

```R
install.package("RColorBrewer")
```



Input requirements
------------------

Most functions take as input a simple data structure, an object of class
`divtable` which is a table of site (rows) by
source (columns) counts, which each cell representing the number of times that
source/species/type was observed at that site.

One method to create a random `divtable`:

```R
n.sites <- 5
n.sources <- 10
n.samples <- 120
## data frame of site-source pairs
t <- data.frame(site = sample(n.sites, n.samples, replace = TRUE),
                source = round(runif(n.samples) * n.sources + 0.5))

## site-by-source matrix
m1 <- do.call(rbind, lapply(split(t, t$site), function(x) table(x$source)))
# this creates a class c("matrix")
m1 <- as.divtable(m1)

## or use xtabs
m2 <- xtabs(data = t)
## this creates a class c("xtabs","table")
m2 <- as.divtable(m2)
```



Membership plots
----------------

One of the first things you may want to do is produce a membership plot.
Membership plots provide a visual representation of the site-by-source table.

**change these to plots from a random matrix**

![membershipPlot output in black and white](membershipPlot_example_BW.png)

If `fill.method = "colour"` and the library `RColorBrewer` is available, then
it is used to pick the colours used for filling groups.

![membershipPlot output in colour](membershipPlot_example_colour.png)

Singleton sources (those that appear just once in just one site) are
distinguished using a white background, while multiton sources (those that
appear multiple times but still in just one site) can be distinguished with a
gray background using the option `distinguish.multiton = TRUE`.  Other options
are provided for controlling labelling of the plot and producing output to PDF
or PostScript files.

Both plots show that if the number of categories requiring colour/fill is
sufficiently large, hatching is combined with colours.  Both plots were
produced with `distinguish.multiton = FALSE`.

Further published examples of membership plots produced by this code or earlier
versions of this code can be seen Scofield _et al._ 2010, 2011, 2012.



Diversity statistics
--------------------

The function `diversity` calculates PMI statistics ([Grivet _et al._
2005](http://dx.doi.org/10.1111/j.1365-294X.2005.02680.x), [Scofield _et al._
2010](http://dx.doi.org/10.1111/j.1365-2745.2010.01649.x), [Scofield _et al._
2011](http://dx.doi.org/10.1007/s00442-010-1828-5)) and dispersal diversity
statistics([Scofield _et al._
2012](http://www.jstor.org/stable/10.1086/668202)) statistics
(<i>q<sub>gg</sub></i>, <i>&alpha;<sub>g</sub></i>, etc.).
For a `divtable` table holding site-by-species counts, the function
calculates PMI and diversity statistics three different ways, each of
which treats possible bias differently and each of which is returned in
an otherwise identical sublist:

* `q` sublist, containing <i>q<sub>gg</sub></i>-based statistics, known to be 
  biased (Grivet _et al._ 2005)

* `q.nielsen`sublist, <i>q<sup>*</sup><sub>gg</sub></i>-based, which apply the
  transformation developed by Nielsen _et al._ (2003) to be unbiased and seem
  to perform well (Scofield _et al._ 2010, Scofield _et al._ 2011, Scofield _et
  al._ 2012).

* `r` sublist, <i>r<sub>gg</sub></i>-based, unbiased but poor performers at low
  sample sizes (Grivet _et al._ 2005, Scofield _et al._ 2012)


Several additional functions are provided.

Function | Description
-------- | -----------
`alphaDiversityTest(tab)` | Test for differences in &alpha; diversity among sites within a single dataset
`alphaContrastTest(tab.a, tab.b)` | Test whether there is a difference in the &alpha; diversity between two datasets
`alphaContrastTest3(tab.a, tab.b, tab.c)` | Test whether there is a difference in the &alpha; diversity among three datasets
`pairwiseMeanTest(tab)` | Test whether mean pairwise divergence/overlap among sites is different from the null espectation
`plotPairwiseMeanTest()` | Plot the list returned from the above test for evaluation
`gammaContrastTest(tab.a, tab.b)` | Test whether there is a difference in the &gamma; diversity between two datasets
`gammaContrastTest3(tab.a, tab.b, tab.c)` | Test whether there is a difference in the &gamma; diversity among three datasets
`gammaAccum(tab)` |  Permute species among sites and accumulate gamma diversity across permutations

An example of the output of `plotAlphaTest()`:

![plotAlphaTest example](plotAlphaTest_example.png)

and of `plotPairwiseMeanTest()`:

![plotPairwiseMeanTest example](plotPairwiseMeanTest_example.png)



### plotPairwiseMatrix.R

Provides a function for plotting pairwise diversity matrices as returned by the
`pmiDiversity()` function, examples of which can be seen in Figure 4A-C of
Scofield _et al._ <I>Am Nat</I>.

`plotPairwiseMatrix()`
: Create a visual plot of pairwise divergence or overlap values as calculated by
`pmiDiversity()`

For example, with `tab` defined as above, plot the divergence matrix based on
_r<sub>gg</sub>_ calculations, labelling the axes "Seed Pool", using the
following code:


````R
pmiD <- pmiDiversity(tab)
plotPairwiseMatrix(pairwise.mat = pmiD$r$divergence.mat,
                   pairwise.mean = pmiD$r$divergence,
                   statistic = "divergence",
                   axis.label = "Seed Pool")
````

![plotPairwiseMatrix example](plotPairwiseMatrix_example.png)


### gammaAccum.R

Provides functions for calculating &gamma; accumulation across sites, and
plotting the result, examples of which can be seen in Figure 4D-F of Scofield
_et al._ <I>Am Nat</I>.  The file `pmiDiversity.R` (see above) is required to be in
the same directory, as it provides functions used here.

A typical workflow using these functions would be:

````R
rga.result <- gammaAccum(tab)
plot(rga.result)
````

#### gammaAccum(tab)

Perform a &gamma; diversity accumulation on the site-by-source data in `tab`.
The result is returned in a list, which may be passed to `plot` to
plot the result.  Several arguments control the method of accumulation and
value of &gamma; calculated.  Only the defaults have been tested; the others were
developed while exploring the data and must be considered experimental.

`tab`
: Site-by-source table, same format as that passed to `pmiDiversity()`

`gamma.method`
: Calculate &gamma; using `"r"` (default), `"q.nielsen"` or `"q"`
method (see paper)

`resample.method`
: `"permute"` (default) or `"bootstrap"`; whether to resample
sites without (`"permute"`) or with (`"bootstrap"`) replacement

`accum.method`
: `"random"` (default) or `"proximity"`.  If `proximity` is
used, then `distance.file` must be supplied

`distance.file`
: A file or data.frame containing three columns of data, with
the header/column names being `pool`, `X`, and `Y`, containing the spatial
locations of the seed pools named in the row names of tab; only used with
`accum.method="proximity"`


#### plotGammaAccum(rga.result)

Create a visual plot of &gamma; accumulation results from `runGammaAccum()`.

![plotGammaAccum example](plotGammaAccum_example.png)

#### Additional functions

The following functions typically won't be used separately, use `runGammaAccum()`
instead.

`gammaAccum()`
: Workhorse function for &gamma; accumulation

`gammaAccumStats()`
: Extracts stats from the result of `gammaAccum()`

`runGammaAccumSimple()`
: Wrapper that runs and then returns stats from `gammaAccum()`


Allelic Diversity
-----------------

These functions calculate allelic alpha, beta and gamma diversity as described by Sork et al. (unpublished), following the same conceptual and statistical principles described in Scofield et al. (2012).  These have been used to calculate the structure of allelic diversity for complete progeny genotypes as well as their decomposed male and female gametes, and to compare alpha and gamma diversity of progeny dispersed relatively short vs. long distances.

### Input requirements

Input begins as a file of genotypes in [GenAlEx](http://biology-assets.anu.edu.au/GenAlEx) format, which is read using the [`readGenalex`] R package (<https://github.com/douglasgscofield/readGenalex>) available via CRAN:

```R
install.packages("readGenalex")
```

### Usage

The augmented `data.frame` returned by `readGenalex()` is then processed into a list of tables, one per locus, using `allele.createTableList()`.  If the true ploidy of the input data does not the ploidy of the GenAlEx file, for example if haploid gametes are represented by a pair of homozygous alleles, then use the `new.ploidy=1` argument to `allele.createTableList()` to reduce the ploidy to its true level.

The list of tables is analysed as a unit by `allele.pmiDiversity()`, and can be passed to one of the contrast functions for testing against another list of tables.

### Workflow

The workflow to calculate basic allelic diversity statistics:

```R
library(readGenalex)
source("allelePmiDiversity.R")
dat <- readGenalex("GenAlEx-format-file-of-genotypes.txt")
gt <- allele.createTableList(dat)
div <- allele.pmiDiversity(gt)
```

For comparing allele diversity between two different samples:

```R
library(readGenalex)
source("allelePmiDiversity.R")
source("alleleDiversityTests.R")
dat1 <- readGenalex("file-of-genotypes-sample-1.txt")
dat2 <- readGenalex("file-of-genotypes-sample-2.txt")
gt1 <- allele.createTableList(dat1)
gt2 <- allele.createTableList(dat2)
alpha.contrast <- allele.alphaContrastTest(gt1, gt2)
gamma.contrast <- allele.gammaContrastTest(gt1, gt2)
```

For calculating and plotting gamma accumulation curves across all loci:

```R
library(readGenalex)
source("allelePmiDiversity.R")
source("alleleGammaAccum.R")
dat <- readGenalex("genotypes.txt")
lst <- allele.createTableList(dat)
allele.rga.result <- allele.runGammaAccum(lst)
plotGammaAccum(allele.rga.result)
```

#### Functions in `allelePmiDiversity.R`

`allele.pmiDiversity()`
: The function calculating diversity for a set of loci.  The single argument is a list produced by `allele.createTableList()`, and it uses the function `allele.pmiDiversitySingleLocus()`.

`allele.createTableList()`
: Take a data.frame of genotypes read by readGenalex(), produce a list of allele count tables used by the other functions.  Each entry of the list is, for each locus, a table of site x allele counts, with row names being the site names, and column names being the names given to the individual alleles.

`allele.pmiDiversitySingleLocus()`
: The single argument is, for a single locus, a table of site &times; allele counts, with row names being the site names, and column names being the names given to the individual alleles.

#### Functions in `alleleDiversityTests.R`

`allele.alphaDiversityTest(lst)`
: Test whether there is a difference in the alpha diversity among patches in an allele diversity dataset, that is, whether &beta; = 1 or &delta; = 0 across a collection of patches at a site (see Sork et al.).

`allele.alphaContrastTest(lst.a, lst.b)`
: Test whether there is a difference in the alpha diversity between two lists of allele diversity datasets.

`allele.gammaContrastTest(lst.a, lst.b)`
: Test whether there is a difference in the gamma diversity between two allele diversity datasets.

#### Functions in `alleleGammaAccum.R`

`allele.runGammaAccum(lst)`
: Perform a gamma diversity accumulation on the site-by-source data in tab.  Several arguments control the method of accumulation and value of gamma calculated.  Other arguments are identical to `gammaAccum()`.  Only the defaults have been tested; the others were developed while exploring the data and must be considered experimental.  The result is returned in a list, which may be passed to `plotGammaAccum()` to plot the result.

* * *

We also provide the datasets `granaries_2002_Qlob` and `granaries_2004_Qlob`,
which include assignments of *Quercus lobata* acorns harvested from acorn
woodpecker granaries in 2002 and 2004 to seed source trees.

* * *

# References

Grivet D, Smouse PE, Sork VL. 2005. A novel approach to an old
problem: tracking dispersed seeds. [_Molecular Ecology_ 14: 3585-3595](http://dx.doi.org/10.1111/j.1365-294X.2005.02680.x).

Nielsen R, Tarpy DR, Reeve HK. 2003. Estimating effective paternity
number in social insects and the effective number of alleles in a population.
[_Molecular Ecology_ 12: 3157-3164](http://dx.doi.org/10.1046/j.1365-294X.2003.01994.x).

Scofield DG, Smouse PE, Karubian J, Sork VL.  2012.  Use of
alpha, beta, and gamma diversity measures to characterize seed dispersal by animals.
[_American Naturalist_ 180: 719-732](http://www.jstor.org/stable/10.1086/668202),
[supplement](http://www.jstor.org/stable/full/10.1086/668202#apa), [data](http://dx.doi.org/10.5061/dryad.40kq7).

Scofield DG, Alfaro VR, Sork VL, Grivet D, Martinez E, Papp J, Pluess AR, Koenig WD, Smouse PE.  2011.  Foraging patterns of acorn woodpeckers (<i>Melanerpes
formicivorus</i>) on valley oak (<i>Quercus lobata</i> N&eacute;e) in two California oak
savanna-woodlands. [_Oecologia_ 166: 187-196](http://dx.doi.org/10.1007/s00442-010-1828-5),
[supplement](http://link.springer.com/content/esm/art:10.1007/s00442-010-1828-5/MediaObjects/442_2010_1828_MOESM1_ESM.doc).

Scofield DG, Sork VL, Smouse PE. 2010. Influence of acorn
woodpecker social behaviour on transport of coast live oak (<i>Quercus agrifolia</i>)
acorns in a southern California oak savanna. [_Journal of Ecology_ 98: 561-572](http://dx.doi.org/10.1111/j.1365-2745.2010.01649.x),
[supplement](http://onlinelibrary.wiley.com/doi/10.1111/j.1365-2745.2010.01649.x/suppinfo).



NEWS
0.9.9001 (development version)
------------------------------

-   createAlleleTables will create a class allele_divtables object from
    genotype data in a class genalex object, and will attempt to convert
    other objects to class genalex using as.genalex from the readGenalex
    package prior to creating the allele tables
-   Diversity tests no longer compare log-likelihood values against an
    analytic Χ2 distribution, which was almost always incorrect
-   The calculation of distances for diversity tests is much more
    efficient in time and space, so tests may be performed on data with
    very high counts, such as diversity of OTUs quantified via read
    counts
-   Diversity tests return class diversity_test, which has print and
    plot methods
-   Gamma accumulation functions return class gamma_accum, which has a
    plot method
-   diversity, alleleDiversityTest, alleleContrastTest, and
    alleleContrastTest3 have methods for divtable and allele_divtables
-   Lists of allele diversity tables have class allele_divtables
-   Diversity site-by-source tables have class divtable, with methods to
    convert from table, xtabs, matrix, and data.frame
-   diversity returns separate lists for diversity values calculated
    using the standard diversity calculations (q), classical corrections
    (r), and Nielsen et al. (2003 Molecular Ecology) corrections
    (q.nielsen).
-   nielsenTransform is an exported function
-   membershipPlot with fill.method = "colour" will use the RColorBrewer
    package if it is available; the default palette is "Dark2" and this
    can be changed with fill.palette. Also, "colour" is a fill.method
    synonym for "color".
-   membershipPlot will on option write the plot to a PDF file; the
    option to write to an EPS file is removed
-   membershipPlot drops the method argument along with the plotting of
    pie charts
-   The datasets granaries_2002_Qlob and granaries_2004_Qlob are
    provided, which include assignments of Quercus lobata acorns
    harvested from acorn woodpecker granaries in 2002 and 2004 to seed
    source trees.
-   Numerous functions have been renamed (e.g., pmiDiversity to
    diversity) and data structures reconfigured (e.g., the return value
    from diversity)

0.1
---

-   Initial version available via Github. This is tagged with 0.1 and a
    tarball is available as a release.



R/membershipPlot.R
#' Plot a series of stacked bars indicating membership
#'
#' Plot a series of stacked bars indicating membership of species across
#' sites, items across groups, or however the contents of \code{tab} should
#' be interpreted.  In the descriptions below we will treat \code{tab} as
#' describing species counts across sites.
#'
#' In the plot, species are ordered bottom to top in order of decreasing total
#' abundance.  Species which are present in more than one site are given a
#' unique colour/hatching combination, with hatching used only if the number of
#' species is greater than the number of readily distinguishable colours,
#' currently set to 8.  Multiton species, with more than one individual but all
#' individual appearing within a single site, are coloured grey when
#' \code{fill.method = "colour"}, and are coloured white when \code{fill.method
#' = "bw"} or \code{distinguish.multiton = FALSE}.  Singleton groups, those with
#' just one member, are coloured white within their respective sites.
#'
#' Examples of membership plots in publications can be found in
#' Scofield \emph{et al}. (2010), Scofield \emph{et al}. (2011) and Scofield
#' \emph{et al}. (2012).
#'
#' @param tab   Object of class \code{\link{divtable}} containing counts of
#' sites (rows) by species membership (columns), or any similar object
#' (\code{table}, \code{matrix}, \code{data.frame}, \code{xtabs}) that
#' can be converted with \code{\link{as.divtable}}.  This argument is
#' normalized to sum to 1 within each site so values may be counts or
#' proportions.
#'
#' @param fill.method   \code{"colour"} (the default, with synonym
#' \code{"color"}), or \code{"bw"}, for the method of colouring different
#' species within the plots.  For \code{"colour"}, no more than eight separate
#' colours are chosen; if there are more than eight species, the
#' colours for lower-frequency groups are modified by hatching lines of
#' varying angles and densities.  For \code{"bw"}, fewer grades of grey are
#' used, together with black-and-white patterns and hatching.
#' For \code{"colour"}, if the package \code{\link{RColorBrewer}} is
#' available, colours will be chosen from the pallete designated by the
#' \code{fill.pallete} option.
#' The aesthetics of colour choice are important for distinguishing among
#' species in membership plots.  If \code{\link{RColorBrewer}} is not
#' available, colours are chosen using \code{\link{rainbow}}.
#'
#' @param fill.palette   For \code{fill.method = "colour"}, if the
#' package \code{\link{RColorBrewer}} is available, use this palette to
#' choose the base colours to use.  Default is \code{"Dark2"}.
#'
#' @param distinguish.multiton   For \code{fill.method = "colour"}, whether
#' to distinguish multiton groups (see Description) with grey rather than
#' the default white colour.  Default is \code{TRUE}.
#'
#' @param file   If specified, produce PDF output to \code{file}, using
#' \code{\link{pdf}}.  Options \code{onefile = FALSE, paper = "special"}
#' are used.
#'
#' @param file.dim   If \code{file} is specified, dimensions of the PDF
#' plot in inches
#'
#' @param xlab,ylab,las,cex   Plot parameters (see \code{\link{par}}
#'
#' @param x.mar.width  Width of the X-margin
#'
#' @param l1,l2,annotate.cex,annotate   First and second lines of annotation
#' to appear above bars in a bar plot, the character expansion factor to use,
#' and the function to place them there.  Alternate annotations may be
#' enabled by redefining \code{annotate}.
#'
#' @param cex.names Character expansion factor for the names plotted below
#' the bars
#'
#' @return No value is returned
#'
#' @references
#'
#' Scofield, D. G.,  Sork, V. L. and Smouse, P. E. (2010) Influence of acorn
#' woodpecker social behaviour on transport of coast live oak
#' (\emph{Quercus agrifolia}) acorns in a southern California oak savanna.
#' \emph{Journal of Ecology} 98:561-572.
#'
#' Scofield, D. G., Alfaro, V. R., Sork, V. L., Grivet, D., Martinez, E.,
#' Papp, J., Pluess, A. R., Koenig, W. D. and Smouse, P. E. (2011) Foraging
#' patterns of acorn woodpeckers (\emph{Melanerpes formicivorus}) on valley
#' oak (\emph{Quercus lobata} N\'{e}e) in two California oak savanna-woodlands.
#' Oecologia 166:187-196.
#'
#' Scofield, D. G., Smouse, P. E., Karubian, J. and Sork, V. L. (2012)
#' Use of alpha, beta and gamma diversity measures to characterize seed
#' dispersal by animals.  \emph{American Naturalist} 180:719-732.
#'
#' @examples
#'
#' ## create table of random membership data
#' n.sites <- 5
#' n.sources <- 10
#' n.samples <- 160
#' ## data frame of site-source pairs
#' set.seed(75333)
#' t <- data.frame(site = sample(n.sites, n.samples, replace = TRUE),
#'                 source = round(runif(n.samples) * n.sources + 0.5))
#' ## create site-by-source table
#' tab <- as.divtable(table(t))
#' membershipPlot(tab)
#'
#' @export
#'
#' @name membershipPlot
#'
NULL

membershipPlot <- function(tab, ...) UseMethod("membershipPlot")



#' @rdname membershipPlot
#'
#' @export
#'
membershipPlot.default <- function(tab, ...)
{
    membershipPlot.divtable(as.divtable(tab), ...)
}



#' @rdname membershipPlot
#'
#' @export
#'
membershipPlot.divtable <- function(tab,
    fill.method = c("colour", "bw", "color"), fill.palette = "Dark2",
    distinguish.multiton = TRUE, xlab = "Site", ylab = "Membership",
    las = 1, x.mar.width = ifelse(las >= 2, 4, 3), cex = 0.7, cex.names = 0.85,
    l1 = NULL, l2 = NULL, annotate.cex = 0.9,
    annotate = function() {
        xl <- seq(0.8, by = 1.3, length.out = max(length(l1), length(l2)))
        if (!is.null(l1)) mtext(l1, at = xl, line = 0.8, cex = annotate.cex, xpd = NA)
        if (!is.null(l2)) mtext(l2, at = xl, line = 0.4, cex = annotate.cex, xpd = NA)
    },
    file = NULL, file.dim = c(5.25, 2),
    ...)
{
    fill.method <- match.arg(fill.method)
    if (fill.method == "color") fill.method <- "colour"

    # data comes in 'backwards' to what we expect when organising for the plot
    tab <- tab[rev(1:nrow(tab)), , drop = FALSE]
    tab <- t(tab)
    # so now types are along the rows, groups along the columns

    # Generate colours: white for singleton types, grayscale for types that
    # appear in a single site, colour for multi-site types.  Sort the types in
    # descending order of number of appearances of type
    fill.colours <- rep("", nrow(tab))
    names(fill.colours) <- rownames(tab)
    n.types <- apply(tab, 1, sum)
    n.types <- rev(sort(n.types))
    tab <- tab[names(n.types), , drop = FALSE]
    # Number of sites in which each type can be found
    n.sites <- apply(tab, 1, function(x) sum(x > 0))
    # Indices into type categories
    singleton.idx <- n.types == 1
    multiton.idx <- n.sites == 1 & n.types > 1
    multisite.idx <- n.sites > 1

    if (fill.method == "colour") {

        fill.colours[singleton.idx] <- "white"
        fill.colours[multiton.idx] <- if (distinguish.multiton)
            gray(seq(from = 0.5, to = 0.8, length = sum(multiton.idx)))
        else "white"
        ang <- dens <- numeric(nrow(tab))
        ang[! multisite.idx] <- 0
        dens[! multisite.idx] <- -1
        fill.colours[! multisite.idx] <- "white"
        n.multisite <- sum(multisite.idx)
        if (n.multisite) {  # we need colours
            n.cols <- min(sum(multisite.idx), 8)
            # if RColorBrewer is available, use it for colours, much nicer,
            colours <- if (requireNamespace("RColorBrewer", quietly = TRUE))
                RColorBrewer::brewer.pal(max(n.cols, 3), fill.palette)
            else rainbow(n.cols)
            lo <- n.multisite - n.cols
            # calculate hatching angles and densities
            ang[multisite.idx] <- c(rep(90, n.cols),
                                    rep(c(45, 75, 105, 135), each = n.cols / 2,
                                        length.out = lo))
            dens[multisite.idx] <- c(rep(-1, n.cols),
                                     rep(c(40, 25), each = n.cols / 2,
                                         length.out = lo))
            fill.colours[multisite.idx] <- c(colours, rep(colours, length.out = lo))
        }
        fill.args <- list(angle = ang, density = dens, col = fill.colours)

    } else if (fill.method == "bw") {

        # White for singleton types and for types that appear in a single site;
        # b+w patterns for multi-site types, following black, gray, very
        # slightly off vertical, 45 degrees right, very slightly off
        # horizontal, 45 degrees left.
        ang <- dens <- numeric(nrow(tab))
        ang[!multisite.idx] <- 0
        dens[!multisite.idx] <- -1
        fill.colours[!multisite.idx] <- "white"
        n.multisite <- sum(multisite.idx)
        if (n.multisite) {
            lo <- sum(multisite.idx) - 2
            ang[multisite.idx] <- c(90, 90, rep(c(90, 45, 135), length.out = lo))
            dens[multisite.idx] <- c(-1, -1, rep(c(40, 25), each = 3,
                                                 length.out = lo))
            fill.colours[multisite.idx] <- c("black", gray(0.5), rep(c(gray(0.5),
                "black"), each = 3, length.out = lo))
        }
        fill.args <- list(angle = ang, density = dens, col = fill.colours)

    }

    # adjust to proportions so they sum to 1
    tab <- scale(tab, scale = apply(tab, 2, sum), center = FALSE)
    if (any(abs(apply(tab, 2, sum) - 1) > (.Machine$double.eps^0.5)))
        stop("didn't scale tab correctly")

    if (! is.null(file)) {
        pdf(file = file, width = file.dim[1], height = file.dim[2],
            onefile = FALSE, paper = "special")
        par(mar = c(x.mar.width, 3, 3, 0), mgp = c(1.9, 0.4, 0),
            bty = "n", xpd = NA, las = las, tcl = -0.25, cex = cex)
    } else {
        par(mar = c(x.mar.width, 4, 3, 1), bty = "n", xpd = NA, las = las)
    }

    tab <- tab[, rev(1:ncol(tab)), drop = FALSE]
    do.call("barplot", c(list(height = tab, space = 0.3, horiz = FALSE),
        cex.names = cex.names, fill.args, ...))
    title(xlab = , ylab = ylab)

    #
    if (! is.null(annotate))
        annotate()

    if (! is.null(file)) dev.off()
}

